### ISSUE TITLE
New feature? Java binding for categorical feature support

### ISSUE THREAD

**Issue by shadyelgewily-slimstock** [dmlc/xgboost#8727]
We are using XGBoost4j using the Java binding and we have a strong appetite for categorical feature support, where splitting is based on partitioning the categorical feature as opposed to one-hot encoding and having XGboost considering each category separately. The release notes for v1.6 states:

"In the future, we will continue to improve categorical data support with new features and
optimizations**. Also, we are looking forward to bringing the feature beyond Python binding,
contributions and feedback are welcomed!** Lastly, as a result of experimental status, the
behavior might be subject to change, especially the default value of related
hyper-parameters."

I'm raising this issue because I'm wondering what the status is of the Java binding for these experimental parameters.

We could contribute to create help develop such a binding, if this is still on the roadmap.

**Comment by trivialfis** [comment]
> Is there already a way to communicate to the native C code which columns in the DMatrix should be considered as categorical, and which as numeric

Yes. As referred in your description https://github.com/dmlc/xgboost/pull/7966 .

>  Is there an appetite at XGboost maintainers to release such a Java binding in a stable version any time in the next 3-6 months say

Yes, that would be 2.0 if all goes well.

> Which components are still required to start using categorical features 

For the Java interface, I think we can already get some small examples running, but haven't been able to prioritize it yet. The `feature_type` and supported `tree_method`s are all it needs. However, my understanding is that most users prefer the scala binding over the java binding and we need to extend the feature info setter/getter to scala and have appropriate integration with the spark estimator interface.
